src/sentinel.rs

```diff
@@ -208,4 +241,133 @@ mod test {
+        }
+    }
+
+#[test]
+    fn threshold_claims_requests_added_with_no_keys() {
+        let mut name_key_pairs = Vec::new();
+        let mut sentinel: Sentinel<TestRequest, TestName> = Sentinel::new(CLAIM_THRESHOLDS, KEY_THRESHOLDS);
+        let name = generate_random_name();
+        let request = TestRequest::new(random::<usize>(), name.clone());
+        let claim = TestClaim { value : random::<usize>() };
+        let serialised_claim = claim.serialise();
+        for _ in 0..CLAIM_THRESHOLDS {
+            let key_pair = crypto::sign::gen_keypair();
+            let signature = crypto::sign::sign_detached(&serialised_claim, &key_pair.1);
+            let climant_name = generate_random_name();
+            name_key_pairs.push((climant_name.clone(), key_pair.0.clone()));
+            match sentinel.add_claim(request.clone(), climant_name, signature.clone(), serialised_claim.clone()) {
+                Some(result) => match result {
+                    AddResult::RequestKeys(source_name) => { assert_eq!(request.get_source(), source_name) },
+                    AddResult::Resolved(_, _) => assert!(false)
+                    },
+                    None => assert!(true)
+                }
+        }
+    }
+
+#[test]
+    fn requests_added_with_various_key_size() {
+        let mut name_key_pairs = Vec::new();
+        let mut sentinel: Sentinel<TestRequest, TestName> = Sentinel::new(CLAIM_THRESHOLDS, KEY_THRESHOLDS);
+        let name = generate_random_name();
+        let request = TestRequest::new(random::<usize>(), name.clone());
+        let claim = TestClaim { value : random::<usize>() };
+        let serialised_claim = claim.serialise();
+        for _ in 0..CLAIM_THRESHOLDS {
+            let key_pair = crypto::sign::gen_keypair();
+            let signature = crypto::sign::sign_detached(&serialised_claim, &key_pair.1);
+            let climant_name = generate_random_name();
+            name_key_pairs.push((climant_name.clone(), key_pair.0.clone()));
+            match sentinel.add_claim(request.clone(), climant_name, signature.clone(), serialised_claim.clone()) {
+                Some(result) => match result {
+                    AddResult::RequestKeys(source_name) => { assert_eq!(request.get_source(), source_name) },
+                    AddResult::Resolved(_, _) => assert!(false)
+                    },
+                    None => assert!(true)
+                }
+        }
+
+        // less than KEY_THRESHOLDS kyes received, should return None
+        for index in 0..KEY_THRESHOLDS - 1 {
+            match sentinel.add_keys(request.clone(), name_key_pairs[index].0.clone(), name_key_pairs.clone()) {
+                Some(_) => assert!(false),
+                None => assert!(true)
+            }
+        }
+
+        // KEY_THRESHOLDS kyes received, should not return none
+        match sentinel.add_keys(request.clone(), name_key_pairs[KEY_THRESHOLDS - 1].0.clone(), name_key_pairs.clone()) {
```

Comment:
KEY_THRESHOLDS-1 is a constant, so `add_keys` will see all the keys as if they came from a single source. For the function to accept the keys, they need to come from KEY_THRESHOLDS of different sources.